Prometheus/TextSerializer.cs

```diff
@@ -49,19 +56,77 @@ public async Task WriteFamilyDeclarationAsync(byte[][] headerLines, Cancellation
-
             var valueAsString = value.ToString(CultureInfo.InvariantCulture);
-
             var numBytes = PrometheusConstants.ExportEncoding
                 .GetBytes(valueAsString, 0, valueAsString.Length, _stringBytesBuffer, 0);
 
             await _stream.Value.WriteAsync(_stringBytesBuffer, 0, numBytes, cancel);
             await _stream.Value.WriteAsync(NewLine, 0, NewLine.Length, cancel);
         }
+
+        /// <summary>
+        /// Creates a metric identifier, with an optional name postfix and an optional extra label to append to the end.
+        /// familyname_postfix{labelkey1="labelvalue1",labelkey2="labelvalue2"}
+        /// Note: Terminates with a SPACE
+        /// </summary>
+        public async Task WriteIdentifierPartAsync(byte[] name, byte[] flatennedLabels, CancellationToken cancel,
+            byte[]? postfix = null, byte[]? extraLabelName = null, byte[]? extraLabelValue = null,
+            byte[]? extraLabelValueOpenMetrics = null)
+        {
+            await _stream.Value.WriteAsync(name, 0, name.Length, cancel);
+            if (postfix != null && postfix.Length > 0)
+            {
+                await _stream.Value.WriteAsync(Underscore, 0, Underscore.Length, cancel);
+                await _stream.Value.WriteAsync(postfix, 0, postfix.Length, cancel);
+            }
+
+            if (flatennedLabels.Length > 0 || (extraLabelName != null && extraLabelValue != null))
+            {
+                await _stream.Value.WriteAsync(LBrace, 0, LBrace.Length, cancel);
+                if (flatennedLabels.Length > 0)
+                {
+                    await _stream.Value.WriteAsync(flatennedLabels, 0, flatennedLabels.Length, cancel);
+                }
+
+                if (extraLabelName != null && extraLabelValue != null)
+                {
+                    if (flatennedLabels.Length > 0)
+                    {
+                        await _stream.Value.WriteAsync(Comma, 0, Comma.Length, cancel);
+                    }
+
+                    await _stream.Value.WriteAsync(extraLabelName, 0, extraLabelName.Length, cancel);
+                    await _stream.Value.WriteAsync(Equal, 0, Equal.Length, cancel);
+                    await _stream.Value.WriteAsync(Quote, 0, Quote.Length, cancel);
+                    await _stream.Value.WriteAsync(extraLabelValue, 0, extraLabelValue.Length, cancel);
+                    await _stream.Value.WriteAsync(Quote, 0, Quote.Length, cancel);
+                }
+                await _stream.Value.WriteAsync(RBraceSp, 0, RBraceSp.Length, cancel);
+            }
+            else
+            {
+                await _stream.Value.WriteAsync(Sp, 0, Sp.Length, cancel);
+            }
+        }
+
+        /// <summary>
+        /// Encode the system variable in regular Prometheus form and also return a OpenMetrics variant, these can be
+        /// the same.
+        /// </summary>
+        internal static Tuple<byte[], byte[]> EncodeSystemLabelValue(double value)
```

Comment:
I would prefer some DTO instead of Tuple, so we clearly name every returned value. Or are there extra considerations that force Tuple to be preferred (e.g. perf)?

Reply:
I added a `CanonicalLabel` struct. It includes all three of the optional fields. I couldn't pass it by ref (async method) so had to add an `IsNonEmpty` method along with a sentinel.